src/rez/utils/request_directives.py

```diff
@@ -0,0 +1,280 @@
+        """Parse arguments from directive syntax string"""
+        raise NotImplementedError
+
+    def to_string(self, args):
+        """Format arguments to directive syntax string"""
+        raise NotImplementedError
+
+    def process(self, range_, version, rank=None):
+        """Process requirement's version range"""
+        raise NotImplementedError
+
+
+class DirectiveHarden(DirectiveBase):
+    """Harden directive request version to specific rank"""
+    @classmethod
+    def name(cls):
+        return "harden"
+
+    def parse(self, arg_string):
+        if arg_string:
+            return [int(arg_string[1:-1].strip())]
+        return []
+
+    def to_string(self, args):
+        if args and args[0]:
+            return "%s(%d)" % (self.name(), args[0])
+        return self.name()
+
+    def process(self, range_, version, rank=None):
+        if rank:
+            version = version.trim(rank)
+        hardened = VersionRange.from_version(version)
+        new_range = range_.intersection(hardened)
+        return new_range
+
+
+# helpers
+#
+
+def parse_directive(request):
+    """Parsing directive requests when creating package
+    """
+    if "//" in request:
+        request_, directive = request.split("//", 1)
+    elif "*" in request:
+        request_, directive = _convert_wildcard_to_directive(request)
+        if not directive:
+            return request
+    else:
+        return request
+
+    # parse directive and save into anonymous inventory
+    _directive_args = directive_manager.parse(directive)
+    directive_manager.loaded.put(_directive_args,
+                                 key=request_,
+                                 anonymous=True)
+
+    return request_
+
+
+def _convert_wildcard_to_directive(request):
+    ranks = dict()
+
+    with dewildcard(request) as deer:
+        req = deer.victim
+
+        def ranking(version, rank_):
+            wild_ver = deer.restore(str(version))
+            ranks[wild_ver] = rank_
+        deer.on_version(ranking)
+
+    cleaned_request = str(req)
+    # do some cleanup
+    cleaned_request = deer.restore(cleaned_request)
+
+    if len(ranks) > 1:
+        # should we support this ?
+        return None, None
+    else:
+        rank = next(iter(ranks.values()))
+        directive = "harden(%d)" % rank
+
+        return cleaned_request, directive
+
+
+def bind_directives(package):
+    """Bind previously parsed directives to each variants
+    """
+    anonymous = directive_manager.loaded.storage(anonymous=True)
+
+    for variant in package.iter_variants():
+        requires = copy(variant.variant_requires)
+        requires += variant.get_requires(build_requires=True,
+                                         private_build_requires=True)
+        requires_str = set(map(str, requires))
+
+        data = {r: v for r, v in anonymous.items() if r in requires_str}
+        directive_manager.loaded.put(data, key=variant)
+
+    directive_manager.loaded.clear(anonymous=True)
```

Comment:
What's the `anonymous` stuff about?

Also, this seems like it may not be threadsafe?

Reply:
The `anonymous` was because, the requirement string must be normalized by the schema and turn into the Requirement object. And at that moment, there seems no way to know which package that requirement string belongs to, hence the *anonymous* state.